### DISCUSSION TITLE
Configure scheduler with a config file

### DISCUSSION THREAD

**Discussion by WendellAdriel** [laravel/framework#45145] [Ideas]
## Introduction

Hello everyone, I was thinking that Laravel could have a way to define the scheduler in two ways:
- The current one using the `Console/Kernel.php`
- Using a scheduler config file in `JSON` or `YAML` format at the root of the application

## How it would work

In this approach, the framework would look first if the `scheduler.json` or `scheduler.yml` file is present in the root and use it, if not it would use the configuration in the `Console/Kernel.php` file.

## Why

With this approach in a CI/CD pipeline we could put this file outside the versioned git, so we could update the scheduler without having to deploy code changes. Even if versioned, it offers a quick and intuitive way to modify the scheduled commands for an application.

## Config file example

```
{
  "task": "my-task", // IT CAN USE COMMAND NAME, COMMAND CLASS PATH OR INVOKABLE CLASS FOR SIMPLE FUNCTIONS
  "cron": "everyMinute", // IT CAN USE CRON EXPRESSION OR ANY SCHEDULE FREQUENCY OPTIONS
  "environments": ["staging", "production"],
  "timezone": "America/New_York",
  "overlap": false,
  "one_server": true,
  "background": true,
  "on_maintenance": true,
  "output": [
    {
      "type": "email", // IT CAN BE "send", "append" OR "email"
      "target": "me@example.com", // IT CAN BE FILEPATH OR EMAIL
      "on_failure": true
    }
  ],
  "ping": {
    "before": "https://example.com",
    "after": "https://example.com",
    "success": "https://example.com",
    "failure": "https://example.com",
  }
}
```

## Observations

A command class could have defined some methods like:
- `when` and `skip` that will be used if they are defined as the **Truth Test Condition**.
- `before`, `after`, `onSuccess` and `onFailure` for **Task Hooks**.
- `pingBefore` and `pingAfter` for conditional ping

IDK if this is something that people would like, but IMO it would be a nice feature and if this is something that people are interested in, I would be really happy to help working on the implementation of this feature.

**Comment by henzeb** [comment]
If that's something we want, we should also think about a way to modify it by code so we can add functionality to our cms systems or api's. Also, storage in database would be nice. As in that case we also backup our configuration.

**Reply by WendellAdriel** [reply]
Oh thanks, those are points we need to take in consideration if this is going to be implemented

**Comment by niladam** [comment]
I think this would be a good idea. I would even go further as to provide/implement different means of 'getting' the schedule (maybe drivers? something like from an API, from a config file, from a json file (which can also be fetched once for example) and so on :)

**Reply by WendellAdriel** [reply]
Oh, that's a great idea!
